### ISSUE TITLE
Traefik is not routing "docker push" operations

### ISSUE THREAD

**Issue by bvis** [containous/traefik#988]
### What version of Traefik are you using (`traefik version`)?
v1.1.2 / camembert
The image I'm using is the 1.1.2-alpine.

### What is your environment & configuration (arguments, toml...)?
I have launched it for working in swarm mode with the parameters:

```
docker \
    service create --name traefik-pub \
    --replicas 3 \
    --reserve-memory 100m \
    --publish 80:80 --publish 443:443 --publish 8080:8080 \
    --network traefik-pub \
    traefik:v1.1.2-alpine \
    --constraints="tag==public" \
    --docker \
    --docker.swarmmode \
    --docker.endpoint=tcp://docker-proxy:2375 \
    --docker.domain="${CLUSTER_DOMAIN}" \
    --docker.watch \
    --entryPoints='Name:https Address::443 TLS' \
    --entryPoints='Name:http Address::80' \
    --acme.entrypoint=https \
    --acme=true \
    --acme.email="myemail@mydomain.com" \
    --acme.ondemand=true \
    --acme.onhostrule=true \
    --acme.storage=/etc/traefik/acme/acme.json \
    --web
```

### What did you do?
I have configured a registry service in my cluster it's running under its own network and I have another service that I'm using to add auth to it. Basically the configuration is something like:

```
docker \
    network create --driver overlay registry

docker \
    service create --name registry \
    --reserve-memory 100m \
    --network registry \
    --publish 5000:5000 \
    -e REGISTRY_STORAGE_DELETE_ENABLED=true \
    registry:2

docker \
  service create --name nginx-registry \
    --network traefik-pub \
    --publish 5001:443 \
    --network registry \
    --container-label com.docker.stack.namespace=ops-infrastructure \
    --label com.docker.stack.namespace=ops-infrastructure \
    --label traefik.tags=public \
    --label traefik.docker.network=traefik-pub \
    --label traefik.backend=registry \
    --label traefik.frontend.entryPoints=https \
    --label traefik.frontend.rule=Host:registry.softonic.one \
    --label traefik.port=80 \
  localhost:5000/basi/nginx-registry:v0.1.0
```
This second nginx is configured to work under the 80 port, which is not publicly accessible and it's routed through traefik, that adds the TLS layer thanks to letsencrypt.
On other hand I've added the same certificates temporally as a test in my nginx image to check if it works directly (mapping an ELB from TCP:443->TCP:5001)

### What did you expect to see?
The behaviour should be the same when accessing using the routes:

- ELB 443 -> Traefik 443 -> Nginx 80 -> Registry 5000
- ELB 443 -> Nginx 443 -> Registry 5000

### What did you see instead?
I've seen that the operations `docker login` and `docker pull` work correctly, but `docker push` does not work when I use traefik.

More precisely this is what I see when I launch the operation without traefik:
```
docker push registry.mydomain.com/my-image:latest
The push refers to a repository [registry.mydomain.com/my-image]
e7525efc7275: Layer already exists
6d7504772167: Pushed
192e9fad2abc: Pushed
36e9226e74f8: Pushed
011b303988d2: Pushed
v0.1.2: digest: sha256:8328c5a194a95b453b0b7b3456861a1456b5700190bafbd6d419cd8d13b08c0f size: 1363
```

And this what I see when traefik is routing the push:
```
docker push registry.mydomain.com/my-image:latest
The push refers to a repository [registry.mydomain.com/my-image]
e7525efc7275: Pushing [==================================================>] 6.144 kB
6d7504772167: Pushing [==================================================>] 4.608 kB
192e9fad2abc: Pushing [==================================================>] 3.584 kB
36e9226e74f8: Pushing [>                                                  ] 506.9 kB/50.1 MB
011b303988d2: Pushing [==================================================>] 4.836 MB
blob upload unknown
```
This is the final output, after many retries and some sleeps between.

In the nginx container logs I can see:

```
10.0.0.14 - myusername [22/Dec/2016:01:12:55 +0000] "POST /v2/my-image/blobs/uploads/ HTTP/1.1" 202 0 "-" "docker/1.12.5 go/go1.6.4 git-commit/7392c3b kernel/4.4.39-moby os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.5 \x5C(darwin\x5C))"
```
But when the same operation was done without traefik I see requests like these as well:
```
10.255.0.2 - myusername [22/Dec/2016:01:28:38 +0000] "PATCH /v2/my-image/blobs/uploads/43050fd4-b313-4a1c-b425-b7aebda549a4?_state=8Fnemdrg3MSDRXXdaKjqSbP2h43z2xUehZj7BAoIsp57Ik5hbWUiOiJiYXNpL25naW54LXJlZ2lzdHJ5IiwiVVVJRCI6IjQzMDUwZmQ0LWIzMTMtNGExYy1iNDI1LWI3YWViZGE1NDlhNCIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxNi0xMi0yMlQwMToyODozOC4xNzgyNzk4NzlaIn0%3D HTTP/1.1" 202 0 "-" "docker/1.12.5 go/go1.6.4 git-commit/7392c3b kernel/4.4.39-moby os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.5 \x5C(darwin\x5C))"
10.255.0.2 - myusername [22/Dec/2016:01:28:39 +0000] "PUT /v2/my-image/blobs/uploads/43050fd4-b313-4a1c-b425-b7aebda549a4?_state=eC26OtCQHA86Qm0VcSfKMEJgtamS0z4pKnBkhc4WOCN7Ik5hbWUiOiJiYXNpL25naW54LXJlZ2lzdHJ5IiwiVVVJRCI6IjQzMDUwZmQ0LWIzMTMtNGExYy1iNDI1LWI3YWViZGE1NDlhNCIsIk9mZnNldCI6NDg4LCJTdGFydGVkQXQiOiIyMDE2LTEyLTIyVDAxOjI4OjM4WiJ9&digest=sha256%3Aaa8a6a9d70674453b1ad55abdd0c10ef95ddbeef80fd7d5641af4dcf46c9420a HTTP/1.1" 201 0 "-" "docker/1.12.5 go/go1.6.4 git-commit/7392c3b kernel/4.4.39-moby os/linux arch/amd64 UpstreamClient(Docker-Client/1.12.5 \x5C(darwin\x5C))"
``` 
Does this mean that traefik is having problems with PATCH and/or PUT operations?

**Comment by trecloux** [comment]
@bvis I tried to reproduce the issue and did not succeed for the moment.
Can you try by simplifying a bit your setup for debugging purpose ? For example without TLS/Acme or without swarm mode ?

**Comment by bvis** [comment]
@trecloux Do you mean that when you configured a registry service being served through traefik the push operations were successful?

**Comment by trecloux** [comment]
@bvis Yes.

**Comment by kleinsasserm** [comment]
No it does not work if there are more registry backends configured. We are running five registry container. As of our tests, it seems like that the docker push command is opening up multiple connections in parallel.  Therefore different types of information are pushed to different backends. Setting stickyness in traefik does not resolv the issue because docker push opens up independend sessions and each one could receive another balancing cookie. After we disabled four of five backends the push works successful.
A stickyness implementation based on ip address would resolv this issue, like "sticky = ipaddress | Cookie"

**Comment by bvis** [comment]
@kleinsasserm probably it was the reason. I've done new tests and now it works. As we have a clue of the fail reason I propose to close this bug and check yours.

**Comment by timoreimann** [comment]
I'm inclined to say that this issue should stay closed.

As @kleinsasserm explained quite well, the problem is that Traefik is lacking the "right" method to do session affinity. #1035 already tracks stickiness based on the IP address, which to my understanding should solve the registry problem and any similar case where a client fans out multiple connections. If a cookie-based approach is desired / needed, a separate issue should be created.

Just my 2 cents.

**Comment by kleinsasserm** [comment]
@timoreimann yes, i've looked at #1035 and that one should address this problem quite well.

**Comment by bvis** [comment]
I can close this task if you want, it's just a preference of problem visibility.

If I close this task, as I'm not the only person affected by this feature lack, some people that are trying to figure out what's the problem with the `push` operation in the registry could open duplicated issues, because #1035 does not say a word about how it affects to the particular usage of the registry or similar services.

Is it there any workaround we could use?

**Comment by timoreimann** [comment]
@bvis We could mention this issue in #1035 and describe how it is solved, hoping for the right keywords to lead folks to the other issue once this one is closed.

Admittedly, I don't know how well this works. Given the registry push scenario is something people run into regularly, I'm also fine with keeping this one open.

**Comment by kleinsasserm** [comment]
Yes that's true. At this point there is no workaround with Traefik because it lacks the ip-stickiness. A possible solution is to use Nginx with TCP streams and ip-address hashing for stickiness as mentioned in #1376. Currently we are running this implementation (Nginx) with five registry containers in swarm.

**Comment by bvis** [comment]
I'll close because there are references here from the other issue. I hope all this won't be so messy. ;)
